src/main/java/io/gravitee/policy/json2xml/JsonToXmlTransformationPolicy.java

```diff
@@ -40,62 +41,81 @@
  */
 public class JsonToXmlTransformationPolicy extends JsonToXmlTransformationPolicyV3 implements Policy {
 
+    private static final String MESSAGE_INVALID_PAYLOAD_FAILURE_KEY = "JSON_INVALID_MESSAGE_PAYLOAD";
+
     public JsonToXmlTransformationPolicy(final JsonToXmlTransformationPolicyConfiguration configuration) {
         super(configuration);
     }
 
+    private static void applyHeaders(HttpHeaders headers, Buffer xmlBuffer) {
+        headers.set(HttpHeaderNames.CONTENT_TYPE, CONTENT_TYPE);
+        headers.set(HttpHeaderNames.CONTENT_LENGTH, Integer.toString(xmlBuffer.length()));
+    }
+
     @Override
     public String id() {
         return "json-xml";
     }
 
     @Override
     public Completable onRequest(final HttpExecutionContext ctx) {
-        return ctx
-            .request()
-            .onBody(
-                bodyUpstream ->
-                    bodyUpstream
-                        .flatMap(buffer -> transformToXml(buffer, CharsetHelper.extractCharset(ctx.request().headers())))
-                        .doOnSuccess(xmlBuffer -> applyHeaders(ctx.request().headers(), xmlBuffer))
-            )
-            .onErrorResumeNext(throwable -> interruptWithBadRequest(ctx, throwable));
+        return ctx.request().onBody(body -> transformBodyToXml(ctx, body, ctx.request().headers(), HttpStatusCode.BAD_REQUEST_400));
     }
 
     @Override
     public Completable onResponse(final HttpExecutionContext ctx) {
         return ctx
             .response()
-            .onBody(
-                bodyUpstream ->
-                    bodyUpstream
-                        .flatMap(buffer -> transformToXml(buffer, CharsetHelper.extractCharset(ctx.response().headers())))
-                        .doOnSuccess(xmlBuffer -> applyHeaders(ctx.response().headers(), xmlBuffer))
-            )
-            .onErrorResumeNext(throwable -> interruptWithBadRequest(ctx, throwable));
+            .onBody(body -> transformBodyToXml(ctx, body, ctx.response().headers(), HttpStatusCode.INTERNAL_SERVER_ERROR_500));
+    }
+
+    private Maybe<Buffer> transformBodyToXml(
+        final HttpExecutionContext ctx,
+        final Maybe<Buffer> bodyUpstream,
+        final HttpHeaders httpHeaders,
+        final int failureHttpCode
+    ) {
+        return bodyUpstream
+            .flatMap(buffer -> transformToXml(buffer, CharsetHelper.extractCharset(httpHeaders)))
+            .doOnSuccess(xmlBuffer -> applyHeaders(httpHeaders, xmlBuffer))
```

Comment:
for the Message part, the `.doOnSuccess(xmlBuffer -> applyHeaders(httpHeaders, xmlBuffer))` is done in `transformMessageToXml`

Do you think we can put header computation in `transformToXml` for this case of a body ?